WooCommerce: Add currentlyEditing state for products and variations.

justinshreve opened the pull request:
Adds currentlyEditing state to both products and variations to track the current object being edited.

To test:

1. Run `make test` and ensure all tests succeed.

allendav commented:
For this PR, I have kinda the same question as here: https://github.com/Automattic/wp-calypso/pull/13104/files#r111511067

justinshreve commented:
@allendav In case you missed the reply, I've responded over at https://github.com/Automattic/wp-calypso/pull/13104#discussion_r111604347.

allendav commented on client/extensions/woocommerce/state/ui/products/edits-reducer.js:
```diff
@@ -29,9 +29,14 @@ function editProductAction( edits, action ) {
 	const prevEdits = edits || {};
 	const bucket = product && isNumber( product.id ) && 'updates' || 'creates';
-	const _array = editProduct( prevEdits[ bucket ], product, data );
+	const _product = product || { id: { index: ( prevEdits[ bucket ] || [] ).length } };
```
The internals of the reducers are starting to be a bit opaque - please consider well-named helper functions before expanding further.

allendav commented on client/extensions/woocommerce/state/ui/products/edits-reducer.js:
```diff
@@ -41,21 +46,25 @@ function editProductAttributeAction( edits, action ) {
 	const bucket = product && isNumber( product.id ) && 'updates' || 'creates';
 	const _attributes = editProductAttribute( attributes, attribute, data );
-	const _array = editProduct( prevEdits[ bucket ], product, { attributes: _attributes } );
+	const _product = product || { id: { index: ( prevEdits[ bucket ] || [] ).length } };
```
Hmmm. In line with my comment above, could this be DRY-ed?

allendav commented on client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js:
```diff
@@ -27,24 +27,36 @@ function editProductVariationAction( edits, action ) {
 			found = true;
-			const _array = editProductVariation( productEdits[ bucket ], variation, data );
-			return { ...productEdits, [ bucket ]: _array };
+			const variationId = variation && variation.id || { index: ( productEdits[ bucket ] || [] ).length };
```
That pattern again :)

justinshreve commented:
@allendav I've added a helpers file under `state/ui` so we can use the same pattern for other reducers. I've moved that code for figuring out the next bucket index, as well as the code for getting the updates or creates bucket, as that is repeated a few times as well. I'm sure other abstractions will make themselves obvious over time.  Does this look OK?

allendav commented:
@justinshreve omg that's perfect :shipit:
